### ISSUE TITLE
process creation time check in _is_postmaster_process is fragile when run in LXC

### ISSUE THREAD

**Issue by mbanck** [zalando/patroni#811]
Hi,

https://github.com/zalando/patroni/blob/534829d617546500d3b40d97fce1e97d09dede9e/patroni/postmaster.py#L45

has 

```
            start_time = int(self._postmaster_pid.get('start_time', 0))
            if start_time and abs(self.create_time() - start_time) > 3:
                logger.info('Process %s is not postmaster, too much difference between PID file start time %s and '
'process start time %s', self.pid, self.create_time(), start_time)
```

I think the `logger.info` is inverted as `start_time` is the pid start time, so should be printed first.

**Comment by CyberDem0n** [comment]
You are absolutely right, self.create_time() and start_time are swapped :(

**Comment by mbanck** [comment]
quick follow-up, I came across this because one of the features fails on ci.debian.net and the culprit seems to be `self.create_time()` being several days in the future (note the swap):

```
2018-09-26 05:20:50,478 INFO: Process 7166 is not postmaster, too much difference between PID file start time 1538286721.13 and process start time 1537939195
```

Any idea what could cause that? virtual environment time skew?

**Comment by CyberDem0n** [comment]
Hmm, this is really weird.
Here we are relying on `psutil` module, which calculates process create time from procfs: https://github.com/giampaolo/psutil/blob/master/psutil/_pslinux.py#L1640-L1648
Either something is wrong in the `/proc/<pid>/stat` or `btime` in the `/proc/stat`.

> virtual environment time skew?

What kind of virtualization do you use?

**Comment by CyberDem0n** [comment]
It seems it is failing on one of the earlier steps.
I'll try to figure out why from the logs.

**Comment by mbanck** [comment]
One other question about that feature:

```
  Scenario: check the basic failover in synchronous mode
[...]
    And I shut down postgres0
[...]

  Scenario: check rejoin of the former master with pg_rewind
    Given I add the table splitbrain to postgres0
    And I start postgres0
```

Did `postgres0` get restarted otherwise in the meantime, or how is adding the table `splitbrain` going to be possible, when it is presumably down (and we start it immediately afterwards)? I also see an immediate shutdown request right after the `CREATE TABLE splitbrain()`, is that some cleanup during start?

**Comment by CyberDem0n** [comment]
```cucumber
  Scenario: check the basic failover in synchronous mode                                                                                
    Given I run patronictl.py pause batman                                                                                              
    Then I receive a response returncode 0                                                                                              
    When I sleep for 2 seconds
    And I shut down postgres0                                                                                                           
    And I run patronictl.py resume batman
```
```
2018-09-26 05:20:18,246 INFO: PAUSE: no action.  i am the leader with the lock                                                          
2018-09-26 05:20:18,255 INFO: No PostgreSQL configuration items changed, nothing to reload.
2018-09-26 05:20:20,231 INFO: Lock owner: postgres0; I am postgres0                                                                     
2018-09-26 05:20:20,241 INFO: PAUSE: no action.  i am the leader with the lock
2018-09-26 05:20:20,728 INFO: Leader key is not deleted and Postgresql is not stopped due paused state   
```

In the `pause` mode it leaves postgres running.

After that it needs to "attach" to the running portgres. In order to identify that this is really a postgres process running under given pid from `postmaster.pid`, Patroni compares process.create_time() and start_time from the `postmaster.pid`.

Since in your case the difference between two numbers is too high, Patroni assumes that this is some other process running with such pid and it is safe to ignore an existing pidfile. It passes this pid in the PG_GRANDPARENT_PID and tries to start postgres.

Postgres fails to start, because in addition to lock on the `postmaster.pid` it is also allocating shared memory and so on.

If you have a better idea how to identify the process running under the given pid - please share it.
Basically it boils down to the following question:
There is a `PGDATA/postmaster.pid` on the filesystem. Is there a posgres running under a given `PGDATA` or not?

**Comment by mbanck** [comment]
First off, thanks for explaining about pause etc. I was confused by "I shut down postgres0" and thought that was supposed to be the postgresql server.

About checking, I had a look what `pg_ctl` does and it sends the pid it finds in `postmaster.pid` the 0 signal, which apparently means "process exists and you can send it signals", after checking against own pid etc.:

https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/bin/pg_ctl/pg_ctl.c#l1257

```
     if (kill(pid, 0) == 0)
```

Right now I can't say how fragile `boot_time` and `psutil.create_time` are but I think you are mostly guarding against some other user reusing the pid? In that case, using `kill()` (if that is portable enough, which might be not the case) could be an alternative.

**Comment by CyberDem0n** [comment]
It is very easy to figure out that the process is alive.
What we need to know, is whether the given process is the postgres which is using our `PGDATA`.

**Comment by mbanck** [comment]
In that case the safest bet seems to be parsing `/proc/[self.pid]/cmdline` for `PGDATA` as patroni has that under its control.

I guess we cannot add/track a creation timestamp during PostmasterProcess.start() as shown by the pause, stop and reattach cycle, where that information would have been lost?

**Comment by ants** [comment]
Chipping in as I'm mostly at fault for the code here. The kill(pid, 0) approach failed reasonably often due to pid assignment on startup starting from the same position and Patroni's background thread got the same pid as postmaster had on previous startup. The choice for start time was made because it seemed less fragile. When manual maintenance is done then postgres could be started by sysadmin. The process name can be different. From normal packages I have seen postmaster and postgres, but somebody could have the binary symlinked as pg96, etc. But if real life provides evidence that start time is less reliable then we should change the approach and just document how sysadmin must manually start postgres for pause/resume to work.

**Comment by mbanck** [comment]
I wasn't suggesting just checking the process name (`postgres` or `postmaster`), but also the rest of the cmdline. But I guess there'd be a problem with parsing `cmdline` for `PGDATA` if the admin starts postgres manually and points -D at the config directory from where `PGDATA` is discerned in `postgresql.conf`, right?

Possibly the heuristic "process name is postgres or postmaster or PGDATA appears in cmdline" could be used?

**Comment by ants** [comment]
That heuristic is going to fail when a postgres instance running on a different PGDATA but as the same user happens to get the same pid. That is a pretty unusual circumstance except for test deployments. But a test system is where this whole thing got started...

Maybe something like this pseudocode

    if start_time_matches:
        return True
    elif cmd_is_postgres_or_postmaster:
        if PGDATA_in_cmd:
            return True
        else:
            logger.warning("Could not determine if process {pid} from postmaster.pid is correct postgres instance. Assuming it is.".format(**data))
            return True
    else:
        logger.info("Process {pid} from postmaster.pid with command line {cmd} does not look like postgres.".format(**data))
        return False

**Comment by mbanck** [comment]
Right, that sounds pretty bulletproof.

**Comment by mbanck** [comment]
One more option a colleague suggested: Run ` SHOW data_directory` and check the result against `PGDATA`. This would proof that there is a postgres running in the `PGDATA` and any other postgres process running in that `PGDATA` would write another pid into `postmaster.pid` so it should be the correct one. 

Not sure whether that would break abstractions in the `PostmasterProcess` class though (no other postgres wire protocol communication seems to happen in that class yet).

**Comment by CyberDem0n** [comment]
It requires opening a connection, but we don't know neither host nor port

**Comment by mbanck** [comment]
OK, in any case, the `cmdline` check in @ants pseudocode above is a no-op; if we only emit a warning then I guess it's easier to just drop it. Or make it fail if it is not there.

**Comment by mbanck** [comment]
I've opened https://github.com/giampaolo/psutil/issues/1344 with a suggestion on how to fix this in psutil, but I think it would be good to be more robust in patroni as well

**Comment by CyberDem0n** [comment]
I spend some time reading on different issues and looking on commits in lxcfs project and it seems that the commit causing such weird behavior was reverted more than 1 year ago: https://github.com/lxc/lxcfs/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aclosed+btime

Am I right?

We can impement a kind of workaround, i.e. compare the difference between the `uptime` returned by `sysinfo` (which is not vitualized) and whatevher is available in the `/proc/uptime`.
If the value from `proc` is smaller than `sysinfo.uptime`, we should substract the difference from `create_time`.

But should we really do that? It will break one day, when lxcfs guys will again virtualize the `btime` in the `/proc/stats` alongside with `SDATE` field in `/proc/<PID>/stat`.

**Comment by mbanck** [comment]
It seems that this is an unfortunate version pick in Debian stable, I am running lxc/lxcfs 2.0.7 which got released in May 2017, so between the two PRs you mention above. I'll also file a Debian bug to possibly get this patched/reverted there.

I think the `create_time` problem should be fixed in `psutil`, and indeed, they already have a `.c` file which runs `sysinfo`, see https://github.com/giampaolo/psutil/blob/master/psutil/_psutil_linux.c#L262 

But it would be good for patroni to not rely on the process creation time, or at least do some other/additional checks as outlined above in case they don't match. Just checking for `create_time` to be in the future of the current time might not be enough depending on how recent the host boot or LXC create are.

**Comment by CyberDem0n** [comment]
It is not really possible to "fix" it reliably. Future changes in the `lxcfs` might break `psutil`.

There is actually a difference between `boot_time` and `uptime`:
* `boot_time` (`btime`), the time when the system was booted, in unix seconds. It is only available it the `/proc/stat`
* `uptime`, the number of seconds since `boot_time`, could be obtained from `/proc/uptime` or `sysinfo`

We can't just replace `boot_time` from `/proc/stat` with `sysinfo.uptime`.

Sure, we can do some additional checks to make sure that this is really a postgres process, but I don't think we could invent something better than was already suggested [here](https://github.com/zalando/patroni/issues/811#issuecomment-424721006)

**Comment by mbanck** [comment]
I did fix this in the lxcfs Debian package in the meantime and this fix is now in the latest stretch point release which went live last weekend, so there is no immediate issue anymore.

Still, I think the strategy as outlined by @ants would be useful to implement, so either close this or maybe tag as enhancement?
